File: src/event_listener.cc
// -*- C++ -*-
// Copyright 2012, Evan Klitzke <evan@eklitzke.org>

#include "./event_listener.h"

#include <boost/scoped_array.hpp>

#include <map>
#include <string>
#include <vector>

#include "./embeddable.h"
#include "./js.h"

using v8::Arguments;
using v8::Context;
using v8::External;
using v8::Handle;
using v8::HandleScope;
using v8::Integer;
using v8::Local;
using v8::Object;
using v8::Script;
using v8::String;
using v8::Undefined;
using v8::Value;

namespace e {
std::vector<Persistent<Object> >& EventListener::CallbackMap(
    const std::string &callback_name,
    bool use_capture) {
  if (use_capture)
    return capture_[callback_name];
  else
    return bubble_[callback_name];
}

bool EventListener::Add(const std::string& callback_name,
                        Local<Object> callback,
                        bool use_capture) {
  std::vector<Persistent<Object> > &callback_list = CallbackMap(callback_name,
                                                                use_capture);
  std::vector<Persistent<Object> >::iterator it;
  for (it = callback_list.begin(); it != callback_list.end(); ++it) {
    if ((*it) == callback) {
      return false;
    }
  }
  Persistent<Object> callback_obj = Persistent<Object>::New(callback);
  callback_list.push_back(callback_obj);
  return true;
}

bool EventListener::Remove(const std::string &callback_name,
                           Local<Object> callback,
                           bool use_capture) {
  std::vector<Persistent<Object> > &callback_list = CallbackMap(callback_name,
                                                                use_capture);
  std::vector<Persistent<Object> >::iterator it;
  for (it = callback_list.begin(); it != callback_list.end(); ++it) {
    if (*it == callback) {
      it->Dispose();
      callback_list.erase(it);
      return true;
    }
  }
  return false;
}

bool EventListener::CallHandler(Handle<Value> h,
                                Handle<Object> this_argument,
                                size_t argc,
                                Handle<Value> argv[]) {
  HandleScope scope;
  if (h->IsObject()) {
    Handle<Object> o = Handle<Object>::Cast(h);
    if (o->IsCallable()) {
      o->CallAsFunction(this_argument, argc, argv);
      return true;
    }

    Handle<String> handle_event_string = String::NewSymbol("handleEvent");
    Handle<Value> handle_event = o->Get(handle_event_string);
    if (handle_event->IsObject()) {
      Handle<Object> handle_event_obj = Handle<Object>::Cast(handle_event);
      if (handle_event_obj->IsCallable()) {
        handle_event_obj->CallAsFunction(this_argument, argc, argv);
        return false;
      }
    }
  }
  return false;
}

void EventListener::Dispatch(const std::string& name) {
  Local<Object> globals = Context::GetCurrent()->Global();
  std::vector<Handle<Value> > arguments;
  Dispatch(name, globals, arguments);
}

void EventListener::Dispatch(const std::string& name,
                             const std::vector<Handle<Value> >& args) {
  Local<Object> globals = Context::GetCurrent()->Global();
  Dispatch(name, globals, args);
}

void EventListener::Dispatch(const std::string &name,
                             Handle<Object> this_argument,
                             const std::vector<Handle<Value> > &arguments) {
  std::vector<Persistent<Object> > &captures = capture_[name];
  std::vector<Persistent<Object> > &bubbles = bubble_[name];

  // build up argc and argv
  const size_t argc = arguments.size();
  boost::scoped_array<Handle<Value> > argv(new Handle<Value>[argc]);
  int i = 0;
  for (auto cit = arguments.begin(); cit != arguments.end(); ++cit) {
    argv[i++] = Handle<Value>(*cit);
  }

  // call all of the capture callbacks
  for (auto it = captures.begin(); it != captures.end(); ++it) {
    CallHandler(*it, this_argument, argc, argv.get());
  }

  // call all of the bubble callbacks
  for (auto it = bubbles.begin(); it != bubbles.end(); ++it) {
    CallHandler(*it, this_argument, argc, argv.get());
  }
}
}

File: js/event_listener.js
// Copyright 2012, Evan Klitzke <evan@eklitzke.org>
//
// An implementation of the event listener interface.

var addListener = function (callbacks, name, callback) {
	callbacks[name] = callbacks[name] || [];
	callbacks[name].push(callback);
};

var removeListener = function (callbacks, name, callback) {
	if (name in callbacks) {
		for (var i = 0; i < callbacks.length; i++) {
			if (callbacks[i] === callback) {
				callbacks.splice(i, 1);
				i--;  // correct for the splice
			}
		}

	}
};

var dispatch = function (obj, name, args) {
	var list = obj[name] || [];
	for (var i = 0; i < list.length; i++) {
		var callback = list[i];
		if (callback.handleEvent) {
			callback.handleEvent.apply(this, args);
		} else {
			callback.apply(this, args);
		}
	}
};

var EventListener = function () {
	this.captures = {};
	this.bubbles = {};
};

// add a callback to the event listener
EventListener.prototype.addEventListener = function (name, callback, use_capture) {
	var list = use_capture ? this.captures : this.bubbles;
	addListener(list, name, callback);
};

// remove a callback from the event listener
EventListener.prototype.removeEventListener = function (name, callback, use_capture) {
	var list = use_capture ? this.captures : this.bubbles;
	removeListener(list, name, callback);
};

// dispatch an event to the event listener
EventListener.prototype.dispatch = function (eventName) {
	var event = {
		'bubbles': true,
		'cancelable': false,
		'currentTarget': eventName,
		'target': eventName,
		'timestamp': (new Date()).valueOf(),
		'type': eventName
	};
	var args = Array.prototype.slice.call(arguments)
	args = args.slice(1);

	// right now the event is pretty useless so we don't actually add it in
	if (false) {
		args.unshift(event);
	}

	dispatch(this.captures, eventName, args);
	dispatch(this.bubbles, eventName, args);
};

exports.EventListener = EventListener;
